### DISCUSSION TITLE
Stuck on the first step… (first run)

### DISCUSSION THREAD

**Discussion by MueJosh** [tphakala/birdnet-go#82] [Q&A]
See edit below
Ive never felt that stupid before 

I cant get birdnet-go to run (first start) 
Ive downloaded the tar file (for raspberry; v0.4.4), unpacked it in my home folder and now what? 
Whats the next step to get it running? 

Raspberry Pi 02W 
Bullseye 64bit lite

Edit:
To clarify my issue a bit: 
Is there a command to start the birdnet-go file? 
I can read the txt file but this doesnt help me. 

Edit 2: 
I've downloaded the wrong .tar folder... arm and amd was too close

Now I can configure everything but that's something for tomorrow.
Still dont know how ro get it running (but i havent configured the settings/config yet, maybe the question will answer itself)

**Comment by JAnslinger** [comment]
Did you figure it out? I am stuck at the same point.

**Comment by JAnslinger** [comment]
Alright, I have indeed made some progress using the wiki entry (https://github.com/tphakala/birdnet-go/wiki/BirdNET%E2%80%90Go-Guide). In the following, I will document my journey so far, including the shell commands, for myself and other newcomers. Maybe it will help someone. I will expand the documentation as I make further progress.

I am using a Raspberry Pi 3+ equipped with the latest 64-bit Lite Version (Bullseye) of Raspberry OS.

### **What I did**
1. I downloaded the current TensorFlow Lite C precomiled library from https://github.com/tphakala/tflite_c/releases for my raspberry pi which is the arm-version
`wget https://github.com/tphakala/tflite_c/releases/download/v2.14.0/tflite_c_v2.14.0_linux_arm64.tar.gz`

2. I extracted this file using:
`tar -xzf tflite_c_v2.14.0_linux_arm64.tar.gz`

3. The extracted file was then copied to the library directory:
`sudo cp libtensorflowlite_c.so /usr/local/lib/`

4. Afterwards, I updated the library cache:
`sudo ldconfig`

5. I created a directory for BirdNET-Go:
`mkdir birdnet-go`

6. Then, I navigated into it and downloaded the latest release of the compiled version:
`cd birdnet-go
wget https://github.com/tphakala/birdnet-go/releases/download/20240215/birdnet-go_v20240215_arm64.tar.gz`

7. This file was also extracted:
`tar -xzf birdnet-go_v20240215_arm64.tar.gz`

Upon checking the contents with `ls`, I noticed that there's also a libtensorflowlite_c.so file included. Therefore, when using ARM, it might be possible to skip steps 1 and 2.

8. Now, the BirdNET-Go executable can be launched with the command:
`./birdnet-go`

9. then I opened the then created config file using 
`nano /home/username/.config/birdnet-go/config.yaml`
I edited 
-the locale according to the [guide](https://github.com/tphakala/birdnet-go/wiki/BirdNET%E2%80%90Go-Guide) 
-as well as longitude and latitude (which I obtained from https://www.latlong.net/)
and saved the file using strg+o and strg+x

10. I then tried to run the realtime analysis
`./birdnet-go realtime`  

but got an error 


> username@raspberry3bplus:~/birdnet-go $ ./birdnet-go realtime
> Read config file: /home/username/.config/birdnet-go/config.yaml
> BirdNET GLOBAL 6K V2.4 FP32 model initialized, using 4 threads of available 4 CPUs
> BirdNET locale not set, using English as default
> Starting BirdNET-Go Analyzer in realtime mode
> Threshold: 0.8, sensitivity: 1, interval: 15
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x9b5698]
> 
> goroutine 1 [running]:
> github.com/tphakala/birdnet-go/internal/analysis.RealtimeAnalysis(0x40001ec240)
>         /home/thakala/go/src/github.com/tphakala/birdnet-go/internal/analysis/realtime.go:47 +0x1f8
> github.com/tphakala/birdnet-go/cmd/realtime.Command.func1(0x40001e6e00?, {0xcca26a?, 0x4?, 0xcca1a2?})
>         /home/thakala/go/src/github.com/tphakala/birdnet-go/cmd/realtime/realtime.go:20 +0x20
> github.com/spf13/cobra.(*Command).execute(0x400011ef00, {0x54f6ba0, 0x0, 0x0})
>         /home/thakala/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:983 +0x840
> github.com/spf13/cobra.(*Command).ExecuteC(0x400011ec00)
>         /home/thakala/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
> github.com/spf13/cobra.(*Command).Execute(0x40000446e8?)
>         /home/thakala/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039 +0x1c
> main.main() 
> 


I'll will open an issue and keep you posted.


### Other aspects, that I was able to solve:
- How do I check whether my microphone works?
1. I checked if it is listed via `arecord -l`
2. Then I recorded a file `arecord -D hw:1,0 -f cd -c1 test-mic.wav`  [the number behind hw must correspond with the number in the listed devices from before]----> STRG+C to stop recording
3. Then I played it using my headphones and `aplay test-mic.wav`

- How to enable I opening the birdnet-go executable from everywhere?
1. Going to the birdnet-go folder and using `pwd` to find out the path
2. Edit your .bashrc file: `nano ~/.bashrc`
3. And add the path from above to the end: `export PATH="$PATH:/home/username/birdnet-go"`
4. Now I can just type birdnet-go

### My open questions to be adressed in future edits of this post:
- How do I remove the bug?

- How do I analyze a single file?
In the readme.txt it is mentioned to use a command like `birdnet-go file birdsound.wav --threshold 0.1` but that won't work. It does not know the command.

- How do I autostart the program?

- How do I install birdnet audio output?
_I found a readme.txt in the doc-folder and opened it `cd doc && nano readme.txt`
Here is some information about the audio configuration. Amongst other things a recommendation to install Pulse audio. This I did : `sudo apt-get install pulseaudio && sudo apt-get install gstreamer1.0-pulseaudio`
I then startet pulseaudio using: `pulseaudio --start`  and checked, whether it started using `ps -C pulseaudio`
Then I edited the config file as recommended in the readme.txt using `sudo nano /etc/pulse/daemon.conf`.
But I still don't know how this is applicable in birdnet._

- How do I access the web platform?

- How do I add mqtt support to collaborate with home assistant

**Reply by cellscycling** [reply]
On a raspberry Pi 5 here, and just debugged my issue with Pangslinger's info.  I was puzzled as my USB mic showed up, it works, the debug output showed nothing but what looked like a working mic, followed the docs, and I was getting no IDs.  This did it for me (though I swear pulseaudio was working before as well):
sudo apt-get install pulseaudio && sudo apt-get install gstreamer1.0-pulseaudio
pulseaudio --start and checked, whether it started using ps -C pulseaudio

Also, the "file" function doesn't seem to exist in the new binaries (February) but it works in the older binaries (December).

**Reply by MueJosh** [reply]
Oh wow, thanks for your detailed answer! 
I reflashed my sd card today and did everything how you described it.

after literally 4h trying to get it working its now outputting stuff🎉 

not what I want but atleast something
> vogel@birdnetoutdoor03:~/birdnet-go $ ./birdnet-go realtime
>Read config file: /home/vogel/.config/birdnet-go/config.yaml
>BirdNET GLOBAL 6K V2.4 FP32 model initialized, using 4 threads of available 4 CPUs
>Starting BirdNET-Go Analyzer in realtime mode
>Threshold: 0.8, sensitivity: 1, interval: 15
>[2024-03-30T00:47:06+01:00] [INFO] ⇨ http server started on [::]:8080
>ALSA lib pcm_dsnoop.c:638:(snd_pcm_dsnoop_open) unable to open slave
>ALSA lib pcm_dsnoop.c:638:(snd_pcm_dsnoop_open) unable to open slave
>ALSA lib pcm_dsnoop.c:638:(snd_pcm_dsnoop_open) unable to open slave
>2024/03/30 00:47:06 Device init failed miniaudio: Failed to open backend device

and after reading your answer for the XXXTimes and installed pulsaudio and gstreamer
this happend:
WOHOO

>vogel@birdnetoutdoor03:~/birdnet-go$ ./birdnet-go realtime
>Read config file: /home/vogel/.config/birdnet-go/config.yaml
>BirdNET GLOBAL 6K V2.4 FP32 model initialized, using 4 threads of available 4 CPUs
>Starting BirdNET-Go Analyzer in realtime mode
>Threshold: 0.8, sensitivity: 1, interval: 15
>[2024-03-30T01:01:14+01:00] [INFO] ⇨ http server started on [::]:8080
>Listening ...
>2024/03/30 01:02:51 New detection: Amsel with confidence: 0.944080
>2024/03/30 01:02:55 No species config found, using default actions for amsel
>2024/03/30 01:04:14 New detection: Schwarzspecht with confidence: 0.815411
>2024/03/30 01:04:16 Updating detection: Schwarzspecht with confidence: 0.835307
>2024/03/30 01:04:19 No species config found, using default actions for schwarzspecht
>2024/03/30 01:04:28 New detection: Amsel with confidence: 0.861347
(playd some sounds on my PC) 

it also outputs this stuff- i dont care for now tho- i need sleep ^^

>not found in $PATH
>2024/03/30 01:29:16 Error generating spectrogram for clips/2024/03/dryocopus_martius_96p_20240330T012400Z.wav: error creating >pectrogram with SoX: SoX binary not found: exec: "sox": executable file not found in $PATH
>2024/03/30 01:29:16 Error generating spectrogram for clips/2024/03/turdus_merula_89p_20240330T012148Z.wav: error creating >spectrogram with SoX: SoX binary not found: exec: "sox": executable file not found in $PATH

The Websever works too, what an relief 
![screenshot_birdnetgo](https://github.com/tphakala/birdnet-go/assets/145150281/50a54374-8658-4a54-a4a4-834445bb6af7)

I let it run for the weekend and see how it performs 
I have 2 BirdNET-Pi running simultaneously with basicly the same settings and mic Location)


If you need any information -let me know.

**Reply by JAnslinger** [reply]
I am happy it works for you at least! :D 
I am still waiting for some insight on my bug. But in the meantime I installed the classic birdnet app, which works just fine. Maybe I'll just go with that. :)

**Reply by MueJosh** [reply]
Thanks ^^

I moved the tensorflow lib from the download that came with birdnetgo to the dedicated folder and not the one you described in step 1 or 2. The one in the repo is 3 month newer I think. 

I recommend the BirdNET-Pi fork from "Nachtzuster"

**Reply by JAnslinger** [reply]
Thanks a lot for the recommendation of the Nachtzuster fork. I just switched and it's much faster! :)

**Reply by isZumpo** [reply]
May I suggest using the docker setup mention ed in the readme. With it you don't have to deal with installing libraries etc. A docker run command will be enough.

**Reply by MueJosh** [reply]
> May I suggest using the docker setup mention ed in the readme. With it you don't have to deal with installing libraries etc. A docker run command will be enough.

Is there a way to get the container running after a reboot without the need of typing a command and have another PC always on?

**Reply by isZumpo** [reply]
Yes there are ways to do that. However, I must warn you that troubleshooting issues etc will require more understanding of docker after this. Though it is not too tricky. 

You can modify the first part of the docker run command to be this:

docker run -d --restart unless-stopped

The -d tells docker to run the container in detatched mode, meaning that it runs outside the context of your shell. If you do "docker ps" after this then you will be able to see the container running. If you then want to stop it you will have to do "docker container stop <id of container from docker ps>" 

The --restart unless-stopped makes the container restart on reboots of your computer and even crashes of the container itself.

In your case it would look like this: 

docker run -d --restart unless-stopped -p 8080:8080 --env ALSA_CARD=1 --device /dev/snd -v $HOME/birdnet-go-volumes/config:/config -v $HOME/birdnet-go-volumes/data:/data -e TZ="Europe/Berlin" ghcr.io/tphakala/birdnet-go:latest

Docker-compose is also a popular tool used for managing containers that I might add a little bit of info in the updated docs about. Could be worth reading up a bit on if you find the topic interesting.

**Reply by MueJosh** [reply]
I will see it as soon as there is trouble ^^ 

Thank you so much for the explanation! 
I already started it last night and it ran just fine, stopping it also worked flawlessly. 


> Docker-compose is also a popular tool used for managing containers that I might add a little bit of info in the updated docs about. Could be worth reading up a bit on if you find the topic interesting. 

I'll check this out on the weekend, thanks 👍

**Comment by tomlut** [comment]
How to fix your spectrogram issue: https://github.com/tphakala/birdnet-go/discussions/78#discussioncomment-8895898

**Reply by MueJosh** [reply]
Thanks! 
That was the issue 😄

**Comment by MueJosh** [comment]
> May I suggest using the docker setup mention ed in the readme. With it you don't have to deal with installing libraries etc. A docker run command will be enough.
@isZumpo 

If i do- i get this: 
>vogel@birdnetoutdoor01:~/birdnet-go $ docker run -ti \
>  -p 8080:8080 \
>  --env ALSA_CARD=1
>  --device /dev/snd \
>  -v /path/to/config:/config \
>  -v /path/to/data:/data \
>  ghcr.io/tphakala/birdnet-go:latest
>"docker run" requires at least 1 argument.
>See 'docker run --help'.
>
>Usage:  docker run [OPTIONS] IMAGE [COMMAND] [ARG...]
>
>Create and run a new container from an image
>-bash: --device: command not found

>vogel@birdnetoutdoor01:~ /birdnet-go $ arecord -l
>**** List of CAPTURE Hardware Devices ****
>card 1: Device [USB Audio Device], device 0: USB Audio [USB Audio]
>  Subdevices: 1/1
>  Subdevice #0: subdevice #0
>vogel@birdnetoutdoor01:~/birdnet-go $
>


Or did I get something wrong of setting this up?
Im completly new to docker! I installed it in `/home/Username/ `
same for birdnet-go `git clone https://github.com/tphakala/birdnet-go`

**Reply by isZumpo** [reply]
To me it looks like you are not running it as one large command. Rather, it appears to be executing every line as a command. Try getting rid of the line breaks, something like this instead:

docker run -ti -p 8080:8080 --env ALSA_CARD=1 --device /dev/snd ghcr.io/tphakala/birdnet-go:latest

Once you have the above command running. Then you will probably want to fill in some values for the volume mounts. Since the data will otherwise be lost on container restarts:

-v /path/to/config:/config
-v /path/to/data:/data

Here the /path/to/config and /path/to/config should be swapped for two absolute paths on your computer where the config and data should be stored. For instance:

-v $HOME/birdnet-go-volumes/config:/config
-v $HOME/birdnet-go-volumes/data:/data

Leaving you with a docker command like this:


docker run -ti -p 8080:8080 --env ALSA_CARD=1 --device /dev/snd -v $HOME/birdnet-go-volumes/config:/config -v $HOME/birdnet-go-volumes/data:/data ghcr.io/tphakala/birdnet-go:latest

Let me know how it goes and we can take it from there. Docker can be a bit tricky if you have not used it before. But once you get used to it, you will never want to go back to running things directly on the computer 😊

**Reply by MueJosh** [reply]
yep, that was the issue for that.

now I get this:

> vogel@birdnetoutdoor01:~/birdnet-go $ docker run -ti -p 8080:8080 --env ALSA_CARD=1 --device /dev/snd ghcr.io/tphakala/birdnet-go:latest
> docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: unable to apply cgroup configuration: unable to start unit "docker-3071d5a09944ddf61a8b5d983829dac305aa33b4e0bb9366c4b990a997f69876.scope" (properties [{Name:Description Value:"libcontainer container 3071d5a09944ddf61a8b5d983829dac305aa33b4e0bb9366c4b990a997f69876"} {Name:Slice Value:"user.slice"} {Name:Delegate Value:true} {Name:PIDs Value:@ au [5494]} {Name:MemoryAccounting Value:true} {Name:CPUAccounting Value:true} {Name:IOAccounting Value:true} {Name:TasksAccounting Value:true} {Name:DefaultDependencies Value:false}]): Interactive authentication required.: unknown.
> 

@ au = without space (i dont want to tag some random user)


I get back to this tomorrow, time to sleep.
Thanks for your help!

**Reply by isZumpo** [reply]
Looks like a sneaky > made its way into the command this time:

ghcr.io/tphakala/>birdnet-go:latest
It should be:
ghcr.io/tphakala/birdnet-go:latest

**Reply by MueJosh** [reply]
Oh, that was me while "pasting" it as a Quote here on Github. 
Its not there in the command line :) 

Edited my answer to the correct copy

good catch tho ^^

**Reply by isZumpo** [reply]
Ah ok I see 😅 I suspect there is something wrong with the docker installation on your computer. Did you follow the official guide?

https://docs.docker.com/engine/install/debian/
I would suggest following the apt repository installation guide.

**Reply by evolsoulx** [reply]
https://github.com/tphakala/birdnet-go/discussions/81 fyi, my discussion thread has a pretty easy 'quickstart' in it if you use docker. i don't know how to use docker, but check out that one, might get you guys where you need.

**Reply by MueJosh** [reply]
> Ah ok I see 😅 I suspect there is something wrong with the docker installation on your computer. Did you follow the official guide?
> 
> https://docs.docker.com/engine/install/debian/ I would suggest following the apt repository installation guide.

YAY! I dont know what went wrong the first time but after reinstalling docker and getting docker for non-root, everything works as.

the species_config.csv should also work now (moved it to config folder)

Now the only issue is the Time, it is 2hours off. 
Probably an easy fix- i just dont know how. 
(the "non-docker - BirdNET-Go" gets the right time)

> 2024/03/31 15:47:37 New detection: Black Redstart with confidence: 0.834612
> 2024/03/31 15:47:44 Updating detection: Black Redstart with confidence: 0.933437


as soon as everything works ill write a small summary for new user (in a new discussion  so its not that confusing to read)

EDIT:
It doesnt recognizes some owls, on my BirdNET-Pi i can set the "Species Occurence Frequency Threshold"
Is there something like this on birdnet-go?
 I dont mean the sigmoid sensitivity.

**Reply by isZumpo** [reply]
Glad to hear that it works! Great job!

Yeah, we probably need some metod of setting the timezone in the docker container. Since the container is not aware of the host's time settings, it probably defaults to GMT-0 right now. I can take a look at it this coming weekend when I have access to a computer again. However, if you are up to try something, this should probably work since we are using debian as the base for our container image:

docker run -ti -p 8080:8080 --env ALSA_CARD=1 --device /dev/snd -e TZ="Europe/Stockholm" ghcr.io/tphakala/birdnet-go:latest

See the:

-e TZ="Europe/Stockholm"

Here you should fill in with your own timezone code. You can find it in this Wikipedia list: 
https://en.m.wikipedia.org/wiki/List_of_tz_database_time_zones

If you try it, please let me know if it works.

It sounds like a great idea for you to write a summary. I have a in progress commit where I hope to improve the documentation on how to get started etc. However, it is a bit hard to get it right since I have a lot of experience using docker, making it hard to see what parts are difficult for those less familiar with it. Then maybe we can combine the two somewhat, making the docs great for everyone!

For the detection of owls, maybe this discussion could be worth taking a look at?
https://github.com/tphakala/birdnet-go/discussions/91

**Reply by MueJosh** [reply]
Thanks!

yep, thats is! 

Thanks for the explenation 
this worked fine:
`docker run -ti -p 8080:8080 --env ALSA_CARD=1 --device /dev/snd -v $HOME/birdnet-go-volumes/config:/config -v $HOME/birdnet-go-volumes/data:/data -e TZ="Europe/Berlin" ghcr.io/tphakala/birdnet-go:latest`




Owls:
It doesnt really work, i have set up like this:

> #Example configuration for species-specific confidence thresholds
> #Format: species, confidence
> #- Species: Localized common name or scientific name (case insensitive)
> #- Confidence value: Range from 0.0 to 1.0 for species reporting.
> #Values exceeding 1.0 are used to explicitly exclude species
> 
> #Exclude non-bird labels (e.g., environmental sounds, human activities)
> Dog,2.0
> Engine,2.0
> Environmental,2.0
> Fireworks,2.0
> Gun,2.0
> Human non-vocal,2.0
> Human vocal,2.0
> Human whistle,2.0
> Noise,2.0
> Power tools,2.0
> Siren,2.0
> 
> #Higher threshold for uncommon species (which are frequently misidentified)
> common kingfisher,0.85
> eurasian moorhen,1.0
> common snipe,1.0
> 
> #Lower threshold for owls (due to rarity or specific interest)
> western barn owl,0.25
> boreal owl,0.25
> eurasian eagle-owl,0.25
> eurasian pygmy owl,0.25
> eurasian scops owl,0.25
> little owl,0.25
> long-eared owl,0.25
> northern hawk-owl,0.25
> short-eared owl,0.25
> snowy owl,0.25
> tawny owl,0.25

and it doesnt do anything.
Threshold / minimum Confidence is set to 0.70
Sigmoid Sensitivity 1.25
GPS is set correctly 
(if its comparing its database to ebird detections this will become problematic since where I live there are eurasian eagle-owl but ebird doesnt know.. so they dont even get detected)

it only detects owls above 0.70 (like every other bird) 
should I comment under #91 for this?

See Pictures below, idk if its supposed to look like this 
(EDIT: ignore the config.yaml and species_config.csv in the /home/Username[Vogel]/ directory)
![Home_USERNAME_](https://github.com/tphakala/birdnet-go/assets/145150281/dcd655f4-01e7-4fbd-93ae-1fed6d4b03d3)
![Home_USERNAME_birdnet-go-volumes_](https://github.com/tphakala/birdnet-go/assets/145150281/2acad453-7371-4678-854c-2f338c8d7c4d)
![Home_USERNAME_birdnet-go-volumes_config_](https://github.com/tphakala/birdnet-go/assets/145150281/f1bab840-57f3-4eee-b321-5ec4d74b6c01)
EDIT_2_: Mhmm, on my fresh install with my own Guide it looks different again...  The birdnet-go folder doesnt even exist... thats something for the next weekend.

I start writing now 👌

**Reply by MueJosh** [reply]
see

#92
